### DISCUSSION TITLE
Maintaining type through a transformation

### DISCUSSION THREAD

**Discussion by RickBradford** [hylang/hy#2315] [General]
Please excuse a basic question from a functional-first programmer, but I am trying to work out how to maintain the type (or class) of a list after I transform it.

So, I have a class called Story, and I create several instances of that in a list.

`(setv storys (readwords "stories.txt"))`

I now update that list using a 'map' function and place it in a new variable.

`(setv updatedstorys (list (map(fn [s] (updater.update s)) storys)))`

If I check the variable 'updatedstorys', the list has the correct number of elements, but the compiler does not recognize that the list contains Story elements, and produces an error.

`(lfor s updatedstorys (print (s.name + "  " + endtime.strftime ("%d-%B" ) + " - " + s.perc) ))`

AttributeError: 'NoneType' object has no attribute 'name'

Any pointers on how to do this correctly would be much appreciated.

**Comment by allison-casey** [comment]
Without knowing the surrounding code, my best guess is that `updater.update` is returning `None`

**Reply by RickBradford** [reply]
Yes, that's what it was - a failure in my understanding of Python rather than Hy.

While I'm trying to sort out this bit of code, could I ask what the correct Hy syntax would be for the Python datetime code: 

```
datetimeobj=datetime.datetime.strptime(s.date,"%Y-%m-%d")
endtime = datetimeobj + datetime.timedelta(days=s.gap)
print (endtime.strftime("%d %B"))
```
Thanks for your help

**Reply by scauligi** [reply]
Translating directly to Hy, it'll look like this:

```
(setv datetimeobj (datetime.datetime.strptime s.date "%Y-%m-%d")
      endtime (+ datetimeobj (datetime.timedelta :days s.gap)))
(print (endtime.strftime "%d %B"))
```

**Reply by RickBradford** [reply]
Many thanks.

What a fine project this is - I recently needed to learn Python after a long time working almost exclusively in F#, and find it to be a very useful but rather dull language. Hy is a much more interesting take on Python.
